In [67]:
import numpy as np
import pandas as pd
import requests
import json
import dotenv
import os
from bs4 import BeautifulSoup

In [68]:
botname = 'ds6600lab2'
version = '0.0'
email = 'cfd6yn@virginia.edu'
useragent = f'{botname}/{version} python-requests/{requests.__version__}'
headers = {'User-Agent': useragent}
headers

{'User-Agent': 'ds6600lab2/0.0 python-requests/2.32.5'}

#### Problem 1 part A:

In [69]:
cand = pd.read_csv('cands22.txt')

ParserError: Error tokenizing data. C error: Expected 12 fields in line 851, saw 13


In [ ]:
cand = pd.read_csv('cands22.txt',
                 on_bad_lines='skip'
)
cand.count

<bound method DataFrame.count of       |2022|  |H0AK00105|  |N00039029|            |Thomas Lamb (3)|  |3|  \
0     |2022|  |H0AL01055|  |N00044245|             |Jerry Carl (R)|  |R|   
1     |2022|  |H0AL01063|  |N00044288|            |Wes Lambert (R)|  |R|   
2     |2022|  |H0AL01097|  |N00044750|         |James Averhart (D)|  |D|   
3     |2022|  |H0AL02087|  |N00030768|            |Martha Roby (R)|  |R|   
4     |2022|  |H0AL02137|  |N00045120|          |Will Dismukes (R)|  |R|   
...      ...          ...          ...                          ...  ...   
8909  |2022|  |S8WI00224|  |N00041929|            |Leah Vukmir (R)|  |R|   
8910  |2022|  |S8WV00119|  |N00041293|  |Paula Jean Swearengin (D)|  |D|   
8911  |2022|  |S8WV00127|  |N00035531|           |Evan Jenkins (R)|  |R|   
8912  |2022|  |S8WV00135|  |N00041474|              |Bo Copley (R)|  |R|   
8913  |2022|  |S8WV00143|  |N00012642|       |Patrick Morrisey (R)|  |R|   

      |AK01|  |    |  | | | |.1 | |.2  |3N| | |.3  
0 

In [ ]:
df.head()

0            1            2                     3    4       5   \
0  |2022|  |H0AK00105|  |N00039029|     |Thomas Lamb (3)|  |3|  |AK01|   
1  |2022|  |H0AL01055|  |N00044245|      |Jerry Carl (R)|  |R|  |AL01|   
2  |2022|  |H0AL01063|  |N00044288|     |Wes Lambert (R)|  |R|  |AL01|   
3  |2022|  |H0AL01097|  |N00044750|  |James Averhart (D)|  |D|  |AL01|   
4  |2022|  |H0AL02087|  |N00030768|     |Martha Roby (R)|  |R|  |AL02|   

       6    7    8    9     10   11  
0  |    |  | |  | |  | |  |3N|  | |  
1  |AL01|  |Y|  |Y|  |I|  |RW|  | |  
2  |    |  | |  | |  | |  |RN|  | |  
3  |    |  | |  | |  | |  |DN|  | |  
4  |    |  | |  | |  | |  |RN|  | |

In [ ]:
df = pd.read_csv('cands22.txt',
                 sep=None,
                 engine='python',
                 header=None,
                 names=range(12),
                 usecols=range(12),
)

df.columns = ['Cycle', 'FECCandID', 'CID', 'FirstLastP', 'Party',
              'DistIDRunFor', 'DistIDCurr', 'CurrCand', 'CycleCand',
              'IncumChall', 'Status', 'PCC']

df = df.apply(lambda col: col.str.replace('|', '', regex=False) if col.dtype == 'object' else col)

df.head()

Cycle  FECCandID        CID          FirstLastP Party DistIDRunFor  \
0  2022  H0AK00105  N00039029     Thomas Lamb (3)     3         AK01   
1  2022  H0AL01055  N00044245      Jerry Carl (R)     R         AL01   
2  2022  H0AL01063  N00044288     Wes Lambert (R)     R         AL01   
3  2022  H0AL01097  N00044750  James Averhart (D)     D         AL01   
4  2022  H0AL02087  N00030768     Martha Roby (R)     R         AL02   

  DistIDCurr CurrCand CycleCand IncumChall Status PCC  
0                                              3N      
1       AL01        Y         Y          I     RW      
2                                              RN      
3                                              DN      
4                                              RN

#### Problem 1 part B:

From reading the file candas22.txt using pd.read_csv, a ParserError shows up on line 851, which has 13 fields instead of 12 (the expected number of fields). on_bad_lines='skip' commands pandas to skip the problematic rows instead of raising an error, and continuing to read the rest of the file. Therefore, a data frame can be resulted succesfully without the rows that failed to parse as we tell pandas to ignore the offending rows.

This is not an appropriate way to load the data if we are trying to analyse the data as a whole because there will be data loss by ignoring these problematic rows and dropping them directly. This way of skipping the problematic or uncommon rows can make it harder to understand the root causes of errors. However, this could be a quick and straight-forward method to have a snapshot of the data in general without crashing into errors. 

#### Problem 2 part A:

Paramaters
{
    "LeagueID": "00"
    "Season": "2015-16",
    "SeasonType": "Regular Season"
}

In [ ]:
url = "https://stats.nba.com/js/data/sportvu/2015/shootingTeamData.json"
r = requests.get(url, headers=headers)
r.json()

{'parameters': {'LeagueID': '00',
  'Season': '2015-16',
  'SeasonType': 'Regular Season'},
 'resource': 'teamtrackingshootingefficiency',
 'resultSets': [{'name': 'TeamTrackingShootingEfficiencyStats',
   'headers': ['TEAM_ID',
    'TEAM_CITY',
    'TEAM_NAME',
    'TEAM_ABBREVIATION',
    'TEAM_CODE',
    'GP',
    'MIN',
    'PTS',
    'PTS_DRIVE',
    'FGP_DRIVE',
    'PTS_CLOSE',
    'FGP_CLOSE',
    'PTS_CATCH_SHOOT',
    'FGP_CATCH_SHOOT',
    'PTS_PULL_UP',
    'FGP_PULL_UP',
    'FGA_DRIVE',
    'FGA_CLOSE',
    'FGA_CATCH_SHOOT',
    'FGA_PULL_UP',
    'EFG_PCT',
    'CFGM',
    'CFGA',
    'CFGP',
    'UFGM',
    'UFGA',
    'UFGP',
    'CFG3M',
    'CFG3A',
    'CFG3P',
    'UFG3M',
    'UFG3A',
    'UFG3P'],
   'rowSet': [['1610612744',
     'Golden State',
     'Warriors',
     'GSW',
     '',
     82,
     48.7,
     114.9,
     14.9,
     0.498,
     16.7,
     0.645,
     33.7,
     0.428,
     21.5,
     0.418,
     11.0,
     11.1,
     28.3,
     21.5,
     0.563,
 

 #### Problem 2 part B:

Reading through the JSON structure of the data, the team-by-team data is located at "data['resultSets'][0]['rowSet']".

#### Problem 2 part C:

In [ ]:
df = pd.json_normalize(data['resultSets'][0], record_path=['rowSet'])
df.shape

(30, 33)

In [ ]:
df.head()

0              1          2    3  4   5     6      7     8      9   \
0  1610612744   Golden State   Warriors  GSW     82  48.7  114.9  14.9  0.498   
1  1610612759    San Antonio      Spurs  SAS     82  48.3  103.5  14.8  0.481   
2  1610612739      Cleveland  Cavaliers  CLE     82  48.7  104.3  16.9  0.481   
3  1610612746    Los Angeles   Clippers  LAC     82  48.6  104.5  15.0  0.497   
4  1610612760  Oklahoma City    Thunder  OKC     82  48.6  110.2  16.1  0.480   

   ...     23    24    25     26   27   28     29    30    31     32  
0  ...  0.478  21.2  42.5  0.497  2.3  6.3  0.363  10.8  25.3  0.429  
1  ...  0.506  18.3  39.8  0.460  0.9  2.6  0.341   6.1  15.9  0.381  
2  ...  0.473  18.2  40.7  0.447  1.7  5.7  0.299   9.0  23.9  0.378  
3  ...  0.480  18.9  42.0  0.450  2.0  6.0  0.334   7.7  20.8  0.373  
4  ...  0.497  17.5  38.7  0.451  1.6  5.1  0.321   6.6  18.6  0.356  

[5 rows x 33 columns]

#### Problem 2 part D:

In [ ]:
df.columns = data['resultSets'][0]['headers']
df.head()

TEAM_ID      TEAM_CITY  TEAM_NAME TEAM_ABBREVIATION TEAM_CODE  GP   MIN  \
0  1610612744   Golden State   Warriors               GSW            82  48.7   
1  1610612759    San Antonio      Spurs               SAS            82  48.3   
2  1610612739      Cleveland  Cavaliers               CLE            82  48.7   
3  1610612746    Los Angeles   Clippers               LAC            82  48.6   
4  1610612760  Oklahoma City    Thunder               OKC            82  48.6   

     PTS  PTS_DRIVE  FGP_DRIVE  ...   CFGP  UFGM  UFGA   UFGP  CFG3M  CFG3A  \
0  114.9       14.9      0.498  ...  0.478  21.2  42.5  0.497    2.3    6.3   
1  103.5       14.8      0.481  ...  0.506  18.3  39.8  0.460    0.9    2.6   
2  104.3       16.9      0.481  ...  0.473  18.2  40.7  0.447    1.7    5.7   
3  104.5       15.0      0.497  ...  0.480  18.9  42.0  0.450    2.0    6.0   
4  110.2       16.1      0.480  ...  0.497  17.5  38.7  0.451    1.6    5.1   

   CFG3P  UFG3M  UFG3A  UFG3P  
0  0.363   10.8   25.3  0.429  
1  0.341    6.1   15.9  0.381  
2  0.299    9.0   23.9  0.378  
3  0.334    7.7   20.8  0.373  
4  0.321    6.6   18.6  0.356  

[5 rows x 33 columns]

#### Problem 2 part E:

In [ ]:
df.to_json("nba_team_data_14-15.json", orient='split', indent=4)
"NBA team data 2014 to 2015 saved as 'nba_team_data_14-15.json'"

"NBA team data 2014 to 2015 saved as 'nba_team_data_14-15.json'"

#### Problem 3 part A:

In [ ]:
url = "http://www.reddit.com/r/popular/top.json"
r = requests.get(url, headers=headers)
"Status Code:", r.status_code

('Status Code:', 200)

In [ ]:
df = pd.DataFrame(data_list, columns=['subreddit', 'title', 'ups', 'created_utc'])
df['created_utc'] = pd.to_datetime(df['created_utc'], unit='s')
df.head(25)

subreddit                                              title  \
0   interestingasfuck  Belgian Malinois dog figured his hooman cheati...   
1            BeAmazed  Photographer shows his POV vs the photos he ta...   
2    nextfuckinglevel                  Anime pose don't look unrealistic   
3         MadeMeSmile                            Roommates till the end!   
4   mildlyinteresting  Pepper that fell behind the fridge dried perfe...   
5          Unexpected              Throwing with your off hand in public   
6               meirl                                              meirl   
7                 law  Stephen Miller threatens to arrest JB Pritzker...   
8       OldSchoolCool  My parents 9 +/- months before I was born. (ar...   
9                pics  The entrance to box seats at a college footbal...   
10  mildlyinfuriating                           Fake it till you make it   
11  interestingasfuck  My brain struggles to comprehend how someone e...   
12        MadeMeSmile                              Wholesome interaction   
13            madlads                                     Madlad mother.   
14         technology  Reagan ad that infuriated the President set to...   
15               pics                  Paddle boarding witches, 10/24/25   
16        MadeMeSmile                              He's their go-to guy.   
17   nextfuckinglevel                        An absolute unit of a horse   
18           facepalm  No more “they’re kinda like Nazis,”  we’ve ent...   
19                aww  Captured my sweet boy at the perfect moment 🥹 ...   
20       TikTokCringe                            NyC Subway Average Days   
21    oddlysatisfying                  A school of fish following a duck   
22     popculturechat  Jameela Jamil says she returned to the UK part...   
23               cats                        I hope bro is getting paid.   
24    MurderedByWords   "When's the next MAGA convention? I can't wait!"   

      ups         created_utc  
0   87685 2025-10-25 07:59:39  
1   74394 2025-10-25 01:11:03  
2   66331 2025-10-24 23:40:11  
3   62791 2025-10-25 11:39:08  
4   59287 2025-10-25 04:10:20  
5   54893 2025-10-25 06:37:10  
6   49570 2025-10-25 04:24:58  
7   48536 2025-10-24 23:48:04  
8   46571 2025-10-25 04:47:09  
9   45197 2025-10-24 22:04:43  
10  42856 2025-10-24 23:39:34  
11  43477 2025-10-25 13:14:20  
12  41979 2025-10-25 02:08:28  
13  41787 2025-10-25 10:55:52  
14  40098 2025-10-24 21:45:16  
15  37296 2025-10-25 12:26:18  
16  35846 2025-10-25 12:19:07  
17  35226 2025-10-25 04:45:12  
18  34460 2025-10-24 21:08:21  
19  34191 2025-10-25 02:26:52  
20  34138 2025-10-25 11:08:51  
21  33265 2025-10-24 22:02:48  
22  33112 2025-10-24 22:05:05  
23  31876 2025-10-25 13:23:45  
24  29509 2025-10-24 23:02:08

#### Problem 4 part A:

In [ ]:
from bs4 import BeautifulSoup
import json

In [ ]:
url = "https://maps.spirithalloween.com/api/getAsyncLocations?template=domain&level=domain"
params = {'template': 'domain',
          'level': 'domain',
          'lat': 40.380028,
          'lng': -97.910156,
          'radius': 1800,
          'limit': 2000
}
r = requests.get(url, params=params)
data = r.json()
json.dumps(list(data.keys()), indent=2)

'[\n  "lat",\n  "lng",\n  "originLat",\n  "originLng",\n  "markers",\n  "maplist",\n  "options"\n]'

#### Problem 4 part B:

In [ ]:
stores = data["markers"]
len(stores)

1535

#### Problem 4 part C:

In [ ]:
store_html_list = [store["info"] for store in stores]
len(store_html_list)

1535

#### Problem 4 part D:

In [ ]:
def extract_address2(html_element):
    soup = BeautifulSoup(html_element, "html.parser")
    div = soup.find("div", class_="tlsmap_popup")

    if not div:
        return None
    div_text = div.string
    if not div_text:
        return None
    div_text = div_text[:-1]

    try:
        info_dict = json.loads(div_text)
    except json.JSONDecodeError:
        return None
    return info_dict("address_2", None)

#### Problem 4 part E:

In [ ]:
address2_list = [extract_address2(html) for html in store_html_list]
callable(extract_address2)

TypeError: 'dict' object is not callable